examples/nlp/nl_image_search.py

```diff
@@ -0,0 +1,588 @@
+"""
+Title: Natural Language Search with a Dual Encoder
+Author: [Khalid Salama](https://www.linkedin.com/in/khalid-salama-24403144/)
+Date created: 2021/01/30
+Last modified: 2021/01/30
+Description: Implementation of a dual encoder model for image search with natural language.
+"""
+
+"""
+## Introduction
+
+The example demonstrates how to build a dual encoder (also known as two-tower) neural network
+model to search for images using natural language. The model is inspired by
+the [CLIP](https://openai.com/blog/clip/)
+approach, introduced by Alec Radford et al. The idea is to train a vision encoder and a text
+encoder jointly to project the representation of images and their captions into the same embedding
+space, such that the caption embeddings are located near the embeddings of the images they describe.
+
+This example requires TensorFlow 2.4 or higher.
+In addition, [TensorFlow Hub](https://www.tensorflow.org/hub)
+and [TensorFlow Text](https://www.tensorflow.org/tutorials/tensorflow_text/intro)
+are required for the BERT model, and [TensorFlow Addons](https://www.tensorflow.org/addons)
+is required for the AdamW optimizer. These libraries can be installed using the
+following command:
+
+```python
+pip install -q -U tensorflow-hub tensorflow-text tensorflow-addons
+```
+"""
+
+"""
+## Setup
+"""
+
+import os
+import collections
+import json
+import numpy as np
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+import tensorflow_hub as hub
+import tensorflow_text as text
```

Comment:
`tensorflow_text` is not really used, since preprocessing from BERT is used directly. Or am I missing something?

Reply:
As a side-note, when importing `tensorflow_text` I'd recommend naming it `tf_text`, since `text` can be pretty confusing